pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,340 @@
+			overhead := node.Status.Capacity[resourceName]
+			overhead.Sub(quantity)
+			overheads[resourceName] = overhead
+		}
+	}
+
+	// Populate overhead metrics
+	if err := c.insertGaugeValues(overheads, node, overheadGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+	// Populate total request metrics
+	if err := c.insertGaugeValues(reqs, node, requestsGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+	// Populate total limits metrics
+	if err := c.insertGaugeValues(limits, node, limitsGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+	// Populate allocatable metrics
+	if err := c.insertGaugeValues(allocatable, node, allocatableGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Controller) insertGaugeValues(resources map[v1.ResourceName]resource.Quantity, node *v1.Node, gaugeVec *prometheus.GaugeVec) error {
+
+	for resourceName, quantity := range resources {
+		resourceTypeName := strings.ReplaceAll(strings.ToLower(string(resourceName)), "-", "_")
+		labels, err := c.generateLabels(node, resourceTypeName)
+		// Register the set of labels that are generated for node
+		nodeNamespacedName := types.NamespacedName{Name: node.Name}
+		c.LabelsMap[nodeNamespacedName] = append(c.LabelsMap[nodeNamespacedName], labels)
+		if err != nil {
+			return fmt.Errorf("generate new labels: %w", err)
+		}
+		gauge, err := gaugeVec.GetMetricWith(labels)
+		if err != nil {
+			return fmt.Errorf("generate new gauge: %w", err)
+		}
+		if resourceName == v1.ResourceCPU {
+			gauge.Set(float64(quantity.MilliValue()))
+		} else {
+			gauge.Set(float64(quantity.Value()))
+		}
+	}
+	return nil
+}
+
+func calculateDaemonSetOverhead(pods []v1.Pod) v1.ResourceList {
+	overheads := v1.ResourceList{}
+	// calculating daemonset overhead
+	daemonSetPods := []v1.Pod{}
+	for _, pod := range pods {
+		if pod.GetOwnerReferences()[0].Kind == "DaemonSet" {
+			daemonSetPods = append(daemonSetPods, pod)
+		}
+	}
+	daemonSetRequests, _ := getPodsTotalRequestsAndLimits(daemonSetPods)
+	addResourceQuantity(daemonSetRequests, overheads)
+	return overheads
+}
+
+// GetPodsTotalRequestsAndLimits calculates the total resource requests and limits for the pods.
+// If pod overhead is non-nil, the pod overhead is added to the
+// total container resource requests and to the total container limits which have a non-zero quantity.
+func getPodsTotalRequestsAndLimits(pods []v1.Pod) (reqs map[v1.ResourceName]resource.Quantity, limits map[v1.ResourceName]resource.Quantity) {
+	reqs, limits = map[v1.ResourceName]resource.Quantity{}, map[v1.ResourceName]resource.Quantity{}
+	for _, pod := range pods {
+		// Excluding pods that are completed or failed
+		if pod.Status.Phase == v1.PodFailed || pod.Status.Phase == v1.PodSucceeded {
```

Comment:
Can you use the pod util function `IsTerminal` here? https://github.com/aws/karpenter/blob/main/pkg/utils/pod/scheduling.go#L31-L33

Reply:
Yup. Good suggestion. Using the helper function now.